File: sorting.cpp
#include <set>
#include <string>
#include <chrono>
#include <iostream>
#include <random>
#include <algorithm>
#include <ips4o.hpp>

#include "cli.hpp"

template<typename SortFunc>
void execute_sort_algorithm(const int n, const int seed, const int iterations, SortFunc sort_func) {
    const auto start_generate = std::chrono::high_resolution_clock::now();

    // generates the same arrays as the python version
    std::mt19937 rng(seed);

    std::vector<int> a(n);
    for (int i = 0; i < n; i++) {
        const uint32_t r = rng();
        a[i] = static_cast<int>(r % static_cast<uint32_t>(n));
    }

    const auto end_generate = std::chrono::high_resolution_clock::now();

    const std::chrono::duration<double> elapsed_generate = end_generate - start_generate;
    std::cout << "The random generation of the array a (size=" << n << ") was done in " << elapsed_generate.count() << " s." << std::endl;

    const auto start_sort = std::chrono::high_resolution_clock::now();

    for (int x = 0; x < iterations; x++) {
        auto b = a;
        sort_func(b);
    }

    const auto end_sort = std::chrono::high_resolution_clock::now();

    const std::chrono::duration<double> elapsed_sort = end_sort - start_sort;
    std::cout << "The array a was sorted in " << elapsed_sort.count() << " s." << std::endl;
}

int main(const int argc, char **argv) {
    std::set<std::string> allowed_value_options;
    std::set<std::string> allowed_literal_options;

    allowed_value_options.insert("-n"); // size of array
    allowed_value_options.insert("-s"); // seed
    allowed_value_options.insert("-i"); // iterations
    allowed_literal_options.insert("sort");
    allowed_literal_options.insert("ips4o");

    const CommandLineArguments parsed_args = parse_args(argc, argv, allowed_value_options, allowed_literal_options, 0);

    if (!parsed_args.success) {
        return -1;
    }

    int n = 1024;
    int seed = 1;
    int iterations = 10;
    for (const auto&[name, value] : parsed_args.value_options) {
        if (name == "-n") {
            n = std::stoi(value);
        }
        if (name == "-s") {
            seed = std::stoi(value);
        }
        if (name == "-i") {
            iterations = std::stoi(value);
        }
    }

    if (parsed_args.literal_options.contains("sort")) {
        execute_sort_algorithm(n, seed, iterations, [](auto& v){std::sort(v.begin(), v.end());});
    }

    if (parsed_args.literal_options.contains("ips4o")) {
        execute_sort_algorithm(n, seed, iterations, [](auto& v){ips4o::sort(v.begin(), v.end());});
    }
}

File: sorting.py
import sys

import numpy as np
import argparse
import time
from python_algorithms.insertion_sort import *
from python_algorithms.selection_sort import *

start_generate = time.perf_counter()

parser = argparse.ArgumentParser()
parser.add_argument("-n", default=1024)
parser.add_argument("--seed", "-s", default=5)
parser.add_argument("--iterations", "-i", default=10)
parser.add_argument("--builtin", action="store_true")
parser.add_argument("--numpy", action="store_true")
parser.add_argument("--insertion_sort", action="store_true")
parser.add_argument("--selection_sort", action="store_true")

args = parser.parse_args()
n = int(args.n)
seed = int(args.seed)
iterations = int(args.iterations)
if sum([args.builtin, args.numpy, args.insertion_sort, args.selection_sort]) != 1:
    print("[Error] Please choose either builtin, numpy, insertion or selection sort.")
    sys.exit()

# generates the same arrays as the c++ version
rng = np.random.RandomState(seed)
a = (rng.randint(0, 2**32, size=n, dtype=np.uint32) % n).astype(np.uint32)
if args.builtin:
    a = a.tolist()

end_generate = time.perf_counter()

print(f"The random generation of the array a (size={n}) was done in {end_generate - start_generate} s.")

start_sort = time.perf_counter()

for _ in range(iterations):
    b = a.copy()
    if args.builtin or args.numpy:
        b.sort()
    elif args.insertion_sort:
        insertion_sort(b)
    elif args.selection_sort:
        selection_sort(b)

end_sort = time.perf_counter()

print(f"The array a was sorted with {'numpy' if args.numpy else 'builtin'} in {end_sort - start_sort} s.")